mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBLeaderElector.java

```diff
@@ -0,0 +1,151 @@
+        this.lockClient = lockClient;
+        this.partitionKey = key;
+    }
+
+    @Override
+    @SuppressWarnings("FutureReturnValueIgnored")
+    public void start() {
+       if(!isLeaderElectorRunning() && shouldLeaderElectorBeRunning.get()) {
+           logger.info("starting leader elector");
+           leaderElector.submit(this::tryToBecomeLeader);
+       }
+    }
+
+    @Override
+    public void shutdown() {
+        logger.info("shutting down");
+        shouldLeaderElectorBeRunning.set(false);
+        if (leaderLock != null) {
+            logger.info("releasing lock");
+            leaderLock.close();
+        }
+        try {
+            logger.info("closing lock client");
+            lockClient.close();
+        } catch (IOException e) {
+            logger.error("error timeout waiting on leader election to terminate executor", e);
+        }
+        if (isLeaderElectorRunning.get()) {
+            leaderElector.shutdownNow();
+        }
+        if (leadershipManager.isLeader()) {
+            logger.info("calling stopBeingLeader this may call exit");
+            // this may call exit and does no shutdown behavior so let's call it last
+            leadershipManager.stopBeingLeader();
+        }
+        logger.info("shutdown complete");
+    }
+
+    public boolean isLeaderElectorRunning() {
+        logger.info("leader running: {}", isLeaderElectorRunning.get());
+        return isLeaderElectorRunning.get();
+    }
+
+
+    /**
+     * This function will attempt to become the leader at the heartbeat interval of the lockClient. If
+     * it becomes the leader it will update the leader data and the thread will stop running.
+     */
+    private void tryToBecomeLeader() {
+        final MasterDescription me = leadershipManager.getDescription();
+        try {
+            logger.info("requesting leadership from {}", me.getHostname());
+            isLeaderElectorRunning.set(true);
+            final Optional<LockItem> optionalLock =
+                    lockClient.tryAcquireLock(
+                            AcquireLockOptions.builder(this.partitionKey)
+                                    .withReplaceData(true)
+                                    .withAcquireReleasedLocksConsistently(true)
+                                    .withData(ByteBuffer.wrap(jsonMapper.writeValueAsBytes(me)))
+                                    .build());
+            if (optionalLock.isPresent()) {
+                leaderLock = optionalLock.get();
+                shouldLeaderElectorBeRunning.set(false);
+                leadershipManager.becomeLeader();
+            }
+        } catch (RuntimeException | InterruptedException | JsonProcessingException e) {
+            logger.error("leader elector task has failed it will restart, if this error is frequent there is likely a problem with DynamoDB based leader election", e);
+        } finally {
+            isLeaderElectorRunning.set(false);
```

Comment:
This block is really about restarting the thread. The function to become leader should keep running until it can become the leader. However, exceptions can occur and if they do we should make a best effort to restart. This is a point where I'm unsure about how to handle metrics in new features. Bringing in the logic for the metrics from upstream felt awkward. In the end I used logging for observability. Metric extension is probably a worthy topic for discussion.